geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,419 @@
+import org.joda.time.{Duration, Instant}
+import org.locationtech.geomesa.filter._
+import org.locationtech.geomesa.kafka.consumer.KafkaConsumerFactory
+import org.locationtech.geomesa.kafka.consumer.KafkaStreamLike.KafkaStreamLikeIterator
+import org.locationtech.geomesa.kafka.consumer.offsets.FindOffset
+import org.locationtech.geomesa.utils.geotools.Conversions._
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes.FR
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter._
+import org.opengis.filter.expression.PropertyName
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+class ReplayKafkaConsumerFeatureSource(entry: ContentEntry,
+                                       replaySFT: SimpleFeatureType,
+                                       liveSFT: SimpleFeatureType,
+                                       topic: String,
+                                       kf: KafkaConsumerFactory,
+                                       replayConfig: ReplayConfig,
+                                       query: Query = null)
+  extends KafkaConsumerFeatureSource(entry, replaySFT, query)
+  with Logging {
+
+  import TimestampFilterSplit.split
+
+  // messages are stored as an array where the most recent is at index 0
+  private[kafka] val messages: Array[GeoMessage] = readMessages()
+
+  override def getReaderForFilter(filter: Filter): FR = {
+
+    val reader: Option[FR] = if (messages.isEmpty) {
+      // no data!
+      None
+    } else {
+      split(filter).flatMap { tfs =>
+        val time = tfs.ts
+        val filter = tfs.filter.getOrElse(Filter.INCLUDE)
+
+        snapshot(time).map(_.getReaderForFilter(filter))
+      }
+    }
+
+    reader.getOrElse(new EmptyFeatureReader[SimpleFeatureType, SimpleFeature](replaySFT))
+  }
+
+  private[kafka] def snapshot(time: Option[Long]): Option[ReplaySnapshotFeatureCache] = {
+
+    val (startTime, startIndex) = time
+      .map(ts => (ts, indexAtTime(ts)))
+      .getOrElse((replayConfig.end.getMillis, Some(0)))
+
+    startIndex.map { si =>
+      val endTime = startTime - replayConfig.readBehind.getMillis
+      snapshot(startTime, si, endTime)
+    }.getOrElse(None)
+  }
+
+  /** @return the index of the most recent [[GeoMessage]] at or before the given ``time``
+    */
+  private[kafka] def indexAtTime(time: Long): Option[Int] = {
+
+    if (replayConfig.isInWindow(time)) {
+      // it doesn't matter what the message is, only the time
+      val key: GeoMessage = new Clear(new Instant(time))
+
+      // reverse ordering for reverse ordered ``messages``
+      val ordering = new Ordering[GeoMessage] {
+        override def compare(x: GeoMessage, y: GeoMessage): Int = y.timestamp.compareTo(x.timestamp)
+      }
+
+      var index = java.util.Arrays.binarySearch(messages, key, ordering)
+
+      if (index < 0) {
+        // no message found at sought time
+        index = -index - 1
```

Comment:
It's not equivalent.  The binary search returns the "index of the search key, if it is contained in the array otherwise, <tt>(-(<i>insertion point</i>) - 1)</tt>".  This calculation is to get the correct inseition point.  I added a comment to clarify.